### PR TITLE
Renewal: failure because loan is not renewable

### PR DESCRIPTION
**Pull request by artem-blazhko**
Renewal: failure because loan is not renewable

## Refs

[FAT-841](https://issues.folio.org/browse/FAT-841)

## Screenshots

image

### CODE REVIEW

File: cypress/integration/loans/renewal-failure.spec.js [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+            cy.deleteInstanceApi(instance.id);
+          });
+        cy.updateCirculationRules({
+          rulesAsText: initialCircRules,
+        });
+        cy.deleteLoanPolicy(LOAN_POLICY_ID);
+      });
+  });
+
+  it('C568 Renewal: failure because loan is not renewable', { tags: [TestType.smoke] }, () => {
```
Comment (dmitriyfolio) [inline cypress/integration/loans/renewal-failure.spec.js:167]: I haven't found checking of "** No Override button displays" from step1 in testrail(

File: cypress/support/api/circulation.js [hunk 2]
```diff
@@ -29,7 +29,12 @@ Cypress.Commands.add('createItemCheckinApi', (data) => {
       id: uuid(),
```
Comment (dmitriyfolio) [inline cypress/support/api/circulation.js:29]: try to move createItemCheckinApi into related fragment and stop to use cy.env. Work through return value please

File: cypress/support/constants.js [hunk 3]
```diff
@@ -28,9 +28,11 @@ export const LOAN_TYPE_NAMES = {
 
 export const MATERIAL_TYPE_NAMES = {
   MICROFORM: 'microform',
+  BOOK: 'book',
 };
 
 export const CY_ENV = {
+  CHECKOUT_ITEM: 'checkoutItem',
```
Comment (dmitriyfolio) [inline cypress/support/constants.js:35]: don't support CY_ENV please. I suggest to work with return value of fragments methods instead

File: cypress/integration/loans/renewal-failure.spec.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+      permissions.loansRenew.gui,
+    ])
+      .then(userProperties => {
+        renewUserData.lastName = userProperties.username;
+        renewUserData.id = userProperties.userId;
+        renewUserData.barcode = userProperties.barcode;
+
+        cy.login(userProperties.username, userProperties.password);
+      });
+    cy.getToken(Cypress.env(CY_ENV.DIKU_LOGIN), Cypress.env(CY_ENV.DIKU_PASSWORD))
```
[the remark this answers is not in the document]
Comment (artem-blazhko) [reply cypress/integration/loans/renewal-failure.spec.js:55]: done

File: cypress/integration/loans/renewal-failure.spec.js [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+          rulesAsText: initialCircRules,
+        });
+        cy.deleteLoanPolicy(LOAN_POLICY_ID);
+      });
+  });
+
+  it('C568 Renewal: failure because loan is not renewable', { tags: [TestType.smoke] }, () => {
+    // reason of usage - https://docs.cypress.io/guides/references/error-messages#Uncaught-exceptions-from-your-application
+    // eslint-disable-next-line consistent-return
+    Cypress.on('uncaught:exception', (err, runnable, promise) => {
```
[the remark this answers is not in the document]
Comment (artem-blazhko) [reply cypress/integration/loans/renewal-failure.spec.js:170]: done

[the remark this answers is not in the document]
Comment (artem-blazhko) [reply cypress/integration/loans/renewal-failure.spec.js:170]: Unfortunately, I do know about the existing Jira ticket. But we have an error in the browser console after clicking on the "Renew" button and as a result the test fails. 

And I see this error in console: 

image

File: cypress/support/constants.js (hunk 3 above)
Comment (artem-blazhko) [reply cypress/support/constants.js:35]: Done

File: cypress/integration/loans/renewal-failure.spec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (artem-blazhko) [reply cypress/integration/loans/renewal-failure.spec.js:168]: Do you mean that _spec.js_ file should contain each step described in testrail???
In other words, should I create all related methods in `support\fragments\loans\renewals.js` file?

image

File: cypress/support/api/circulation.js (hunk 2 above)
Comment (artem-blazhko) [reply cypress/support/api/circulation.js:29]: You suggest creating a separate method for callback of `after()` block. But as I see from the code perspective we do not use that approach and directly write the code in `after()` block

File: cypress/integration/loans/renewal-failure.spec.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+          items: [[{
+            barcode: itemData.barcode,
+            status: { name: 'Available' },
+            permanentLoanType: { id: Cypress.env(CY_ENV.LOAN_TYPES)[0].id },
+            materialType: { id: materialTypeId },
+          }]],
+        });
+      })
+      // create loan policy
+      .then(() => {
```
[the remark this answers is not in the document]
Comment (artem-blazhko) [reply cypress/integration/loans/renewal-failure.spec.js:101]: Should I do the same for other steps (creating circulation rule, getting tokens, checkout items) in `before()` block?

File: cypress/integration/loans/renewal-failure.spec.js (hunk 1 above)
Comment (artem-blazhko) [reply cypress/integration/loans/renewal-failure.spec.js:168]: Yes, It doesn't exist. I have some difficulties with it. 
User doesn't have permission to see 'Override' button but button exists. If I create the user with the same permissions manually button has correct behavior and doesn't exist. I do not know why it happens and how to handle it. 
And I will appreciate it if someone can help with it or suggest something.

Also, I asked about it in QA channel:
image

File: cypress/integration/loans/renewal-failure.spec.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (dmitriyfolio) [reply cypress/integration/loans/renewal-failure.spec.js:101]: For example, related with "Loan policy that says item is loanable, but not renewable". Can you move preparing of this loan policy into something like createLoanableNotRenewableLoanPolicyViaAPI() into loan policies fragment? This method will contain all steps related with this loan creation, including all related objects management.

File: cypress/integration/loans/renewal-failure.spec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmitriyfolio) [reply cypress/integration/loans/renewal-failure.spec.js:168]: I mean that test in code should be logically crossed with test in testrail. It helps in test design and code review.

[the remark this answers is not in the document]
Comment (dmitriyfolio) [reply cypress/integration/loans/renewal-failure.spec.js:168]: for example - I prefer to keep navigation steps in test level.

File: cypress/support/api/circulation.js (hunk 2 above)
Comment (dmitriyfolio) [reply cypress/support/api/circulation.js:29]: let's discuss this in call by voice

File: cypress/integration/loans/renewal-failure.spec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (artem-blazhko) [reply cypress/integration/loans/renewal-failure.spec.js:161]: Done

File: cypress/support/api/circulation.js (hunk 2 above)
Comment (artem-blazhko) [reply cypress/support/api/circulation.js:29]: done

File: cypress/support/fragments/loans/renewals.js [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+export default {
+  renewWithoutOverrideAccess(loanId, userId, itemData) {
+    cy.visit(generateInitialLink(userId, loanId));
+
+    checkLoansPage();
+
+    checkModalTable(headers.renewConfirmation, itemData);
+
+    // todo: Uncomment button check after permission issue fix
```
[the remark this answers is not in the document]
Comment (artem-blazhko) [reply cypress/support/fragments/loans/renewals.js:125]: That ticket is not related. 
As we discussed this permission issue happens only with testing. But everything works well if we work manually.

File: cypress/integration/loans/renewal-failure.spec.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+import checkoutActions from '../../support/fragments/checkout/checkout';
+import checkinActions from '../../support/fragments/check-in-actions/checkInActions';
+
+describe('Renewal', () => {
+  let materialTypeId;
+  let loanId;
+  let servicePointId;
+  let initialCircRules;
+  const firstName = 'testPermFirst';
+  const renewUserData = {
```
[the remark this answers is not in the document]
Comment (artem-blazhko) [reply cypress/integration/loans/renewal-failure.spec.js:23]: Fixed but did not move to fragment. I need that objects outside the fragment

File: cypress/support/api/circulation.js (hunk 2 above)
Comment (artem-blazhko) [reply cypress/support/api/circulation.js:29]: It has related fragments. 
Or do you mean that we should have separated fragments that contain ONLY api calls?

Comment (dmitriyfolio) [reply cypress/support/api/circulation.js:29]: "It has related fragments." - could you move these methods into related existing fragements?
"do you mean that we should have separated fragments that contain ONLY api calls?" - no. Idea - bisuness related with GUI fragments-> fragments has related with business GUI and api methods

Comment (artem-blazhko) [reply cypress/support/api/circulation.js:29]: But I already moved `createItemCheckinApi` to already existing `fragments/check-in-actions/checkInActions.js` file.
And `checkoutItemApi` to `fragments/checkout/checkout.js` file.

Comment (dmitriyfolio) [reply cypress/support/api/circulation.js:29]: let's remove cypress/support/api/circulation.js in this case

Comment (artem-blazhko) [reply cypress/support/api/circulation.js:29]: Done

File: cypress/integration/loans/renewal-failure.spec.js (hunk 1 above)
Comment (dmitriyfolio) [reply cypress/integration/loans/renewal-failure.spec.js:167]: can't see related todo

**Review by dmitriyfolio — APPROVED** [review]

**Pull request merged**